### DISCUSSION TITLE
Graph Encoder Decoder

### DISCUSSION THREAD

**Discussion by yogeshhk** [pyg-team/pytorch_geometric#2826] [Ideas]
Wish to build encoder-decoder network which accepts graphs as input as well as outputs. Both can have different topology. Similar to Machine Translation, where input-language-sentence-graph can have different topology than input-language-sentence-graph.

Is there any Graph Neural Network architecture which can learn this input-to-output-graph transformation in end-to-end with supervised manner. Dataset having such input-output graphs pairs is available.

How to build such encoder decode network? Please note that as both, input and output are different, this can not be AutoEncoder.

More details on the original problem at: https://github.com/yogeshhk/MidcurveNN

**Comment by rusty1s** [comment]
This is indeed a tricky problem due to the permutation invariance requirement. In general, if your nodes do not change, i.e., you simply want to learn a new graph-structure, you can follow a similar procedure as in Auto-Encoders or Link prediction examples by just training against a different ground-truth.

If you also want to learn/generate new nodes, I think you may need some more sophisticated techniques from recent graph generation proposals, see, e.g., [here](https://github.com/rusty1s/pytorch_geometric/discussions/2614).
You can also think about extending `MidcurveNN` in a way that it learns, in addition to the set of output points, its graph connectivity.

**Reply by yogeshhk** [reply]
Thanks @rusty1s for the response. Yes, Midcurve is typically offset of some of the original edge, so learning input and output connectivity is essential.

Just came across http://proceedings.mlr.press/v119/shi20d/shi20d.pdf to see if it has some parallels with my topic.
